File: app/src/main/java/com/networks/testapplication/utils/ObservableHorizontalScrollView.java
package com.networks.testapplication.utils;

import android.content.Context;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.widget.HorizontalScrollView;

public class ObservableHorizontalScrollView extends HorizontalScrollView {
    public interface OnScrollListener {
        public void onScrollChanged(ObservableHorizontalScrollView scrollView, int x, int y, int oldX, int oldY);
        public void onEndScroll(ObservableHorizontalScrollView scrollView);
    }

    private boolean mIsScrolling;
    private boolean mIsTouching;
    private Runnable mScrollingRunnable;
    private OnScrollListener mOnScrollListener;

    public ObservableHorizontalScrollView(Context context) {
        this(context, null, 0);
    }

    public ObservableHorizontalScrollView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ObservableHorizontalScrollView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    public boolean onTouchEvent(MotionEvent ev) {
        int action = ev.getAction();

        if (action == MotionEvent.ACTION_MOVE) {
            mIsTouching = true;
            mIsScrolling = true;
        } else if (action == MotionEvent.ACTION_UP) {
            if (mIsTouching && !mIsScrolling) {
                if (mOnScrollListener != null) {
                    mOnScrollListener.onEndScroll(this);
                }
            }

            mIsTouching = false;
        }

        return super.onTouchEvent(ev);
    }

    @Override
    protected void onScrollChanged(int x, int y, int oldX, int oldY) {
        super.onScrollChanged(x, y, oldX, oldY);

        if (Math.abs(oldX - x) > 0) {
            if (mScrollingRunnable != null) {
                removeCallbacks(mScrollingRunnable);
            }

            mScrollingRunnable = new Runnable() {
                public void run() {
                    if (mIsScrolling && !mIsTouching) {
                        if (mOnScrollListener != null) {
                            mOnScrollListener.onEndScroll(ObservableHorizontalScrollView.this);
                        }
                    }

                    mIsScrolling = false;
                    mScrollingRunnable = null;
                }
            };

            postDelayed(mScrollingRunnable, 200);
        }

        if (mOnScrollListener != null) {
            mOnScrollListener.onScrollChanged(this, x, y, oldX, oldY);
        }
    }

    public OnScrollListener getOnScrollListener() {
        return mOnScrollListener;
    }

    public void setOnScrollListener(OnScrollListener mOnEndScrollListener) {
        this.mOnScrollListener = mOnEndScrollListener;
    }

}

File: app/src/main/java/com/networks/testapplication/utils/ObservableHorizontalScrollView.kt
package com.networks.testapplication.utils

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.widget.HorizontalScrollView

class ObservableHorizontalScrollView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : HorizontalScrollView(context, attrs, defStyle) {
    interface OnScrollListener {
        fun onScrollChanged(
            scrollView: ObservableHorizontalScrollView?,
            x: Int,
            y: Int,
            oldX: Int,
            oldY: Int
        )

        fun onEndScroll(scrollView: ObservableHorizontalScrollView?)
    }

    private var mIsScrolling = false
    private var mIsTouching = false
    private var mScrollingRunnable: Runnable? = null
    var onScrollListener: OnScrollListener? =
        null
    private var mIsScrollable = true
    fun setIsScrollable(mIsScrollable: Boolean) {
        this.mIsScrollable = mIsScrollable
    }

    override fun onTouchEvent(ev: MotionEvent): Boolean {
        if (!mIsScrollable) return false
        val action = ev.action
        if (action == MotionEvent.ACTION_MOVE) {
            mIsTouching = true
            mIsScrolling = true
        } else if (action == MotionEvent.ACTION_UP) {
            if (mIsTouching && !mIsScrolling) {
                if (onScrollListener != null) {
                    onScrollListener!!.onEndScroll(this)
                }
            }
            mIsTouching = false
        }
        return super.onTouchEvent(ev)
    }

    override fun onScrollChanged(x: Int, y: Int, oldX: Int, oldY: Int) {
        super.onScrollChanged(x, y, oldX, oldY)
        if (Math.abs(oldX - x) > 0) {
            if (mScrollingRunnable != null) {
                removeCallbacks(mScrollingRunnable)
            }
            mScrollingRunnable = Runnable {
                if (mIsScrolling && !mIsTouching) {
                    if (onScrollListener != null) {
                        onScrollListener!!.onEndScroll(this@ObservableHorizontalScrollView)
                    }
                }
                mIsScrolling = false
                mScrollingRunnable = null
            }
            postDelayed(mScrollingRunnable, 200)
        }
        if (onScrollListener != null) {
            onScrollListener!!.onScrollChanged(this, x, y, oldX, oldY)
        }
    }

}